### DISCUSSION TITLE
Popen native or C import

### DISCUSSION THREAD

**Discussion by john9631** [exaloop/codon#458] [Q&A] [answered]
I checked the codebase and there is no native subprocess module.  Os defines def system(cmd: str) -> int but not popen.

I don't want to add the weight of importing pOpen from python's subprocess so decided to try using the C popen.   A simplified version of my code looks like:

```
from C import popen(str, str) -> cobj
from C import fgets(cobj, int, cobj)

pp = popen("ls -l", "r")
print(f"pp is: {pp}, with type: {type(pp)}")
```

which returns

`pp is: 0x0, with type: partial(__new__[...,...])`

where this seems to be a pointer to type byte ( `expected type 'Ptr[byte]'` ).  Does anyone know how I should be handling popen's return object in codon?  My goal is to be able to manipulate the resulting string or array of strings.

**Comment by arshajii** [comment, marked as answer]
Hi @john9631 -- `popen` should actually take two `cobj`s (which correspond to the `char *` arguments in C). Here is the modified code which also calls `fgets`:

``` python
from C import popen(cobj, cobj) -> cobj
from C import fgets(cobj, i32, cobj) -> cobj

pp = popen("ls -l".c_str(), "r".c_str())

n = 1000
p = cobj(n)
fgets(p, i32(n), pp)
print('read:', str.from_ptr(p))
```

**Reply by john9631** [reply]
Thanks.

I extended it to my real problem and it worked perfectly.  You've done a lovely job of keeping this as simple as possible.
